src/lib/flow/sol-flow-js.c

```diff
@@ -0,0 +1,1422 @@
+    duk_push_int(mdata->duk_ctx, value.step);
+    duk_put_prop_string(mdata->duk_ctx, obj_idx, "step");
+
+    if (duk_pcall_method(mdata->duk_ctx, 1) != DUK_EXEC_SUCCESS) {
+        SOL_ERR("Javascript %s process() function error: %s\n",
+            get_in_port_name((struct flow_js_type *)node->type, port), duk_safe_to_string(mdata->duk_ctx, -1));
+        duk_pop_2(mdata->duk_ctx); /* process() result and global_stash */
+        return -1;
+    }
+
+    duk_pop_2(mdata->duk_ctx); /* process() result and global_stash */
+
+    return 0;
+}
+
+static int
+rgb_process(struct sol_flow_node *node, void *data, uint16_t port, uint16_t conn_id,
+    const struct sol_flow_packet *packet)
+{
+    const struct flow_js_data *mdata = (struct flow_js_data *)data;
+    struct sol_rgb value;
+    duk_idx_t obj_idx;
+    int r;
+
+    r = sol_flow_packet_get_rgb(packet, &value);
+    SOL_INT_CHECK(r, < 0, r);
+
+    duk_push_global_stash(mdata->duk_ctx);
+
+    if (!duk_get_prop_index(mdata->duk_ctx, -1, port * PORTS_IN_METHODS_LENGTH + PORTS_IN_PROCESS_INDEX)) {
+        SOL_ERR("Couldn't handle '%s' process().", get_in_port_name((struct flow_js_type *)node->type, port));
+        duk_pop_2(mdata->duk_ctx); /* get_prop() value and global_stash */
+        return -1;
+    }
+
+    if (duk_is_null_or_undefined(mdata->duk_ctx, -1)) {
+        SOL_WRN("'%s' process() callback not implemented in javascript, ignoring incoming packets for this port",
+            get_in_port_name((struct flow_js_type *)node->type, port));
+        duk_pop_2(mdata->duk_ctx); /* get_prop() value and global_stash */
+        return 0;
+    }
+
+    /* In order to use 'node' object as 'this' binding. */
+    duk_dup(mdata->duk_ctx, -3);
+
+    obj_idx = duk_push_object(mdata->duk_ctx);
+    duk_push_int(mdata->duk_ctx, value.red);
+    duk_put_prop_string(mdata->duk_ctx, obj_idx, "red");
+    duk_push_int(mdata->duk_ctx, value.green);
+    duk_put_prop_string(mdata->duk_ctx, obj_idx, "green");
+    duk_push_int(mdata->duk_ctx, value.blue);
+    duk_put_prop_string(mdata->duk_ctx, obj_idx, "blue");
+    duk_push_int(mdata->duk_ctx, value.red_max);
+    duk_put_prop_string(mdata->duk_ctx, obj_idx, "red_max");
+    duk_push_int(mdata->duk_ctx, value.green_max);
+    duk_put_prop_string(mdata->duk_ctx, obj_idx, "green_max");
+    duk_push_int(mdata->duk_ctx, value.blue_max);
+    duk_put_prop_string(mdata->duk_ctx, obj_idx, "blue_max");
+
+    if (duk_pcall_method(mdata->duk_ctx, 1) != DUK_EXEC_SUCCESS) {
+        SOL_ERR("Javascript %s process() function error: %s\n",
+            get_in_port_name((struct flow_js_type *)node->type, port), duk_safe_to_string(mdata->duk_ctx, -1));
+        duk_pop_2(mdata->duk_ctx); /* process() result and global_stash */
+        return -1;
+    }
+
+    duk_pop_2(mdata->duk_ctx); /* process() result and global_stash */
+
+    return 0;
+}
+
+static int
+string_process(struct sol_flow_node *node, void *data, uint16_t port,
```

Comment:
These process functions have a lot of common boilerplate; can't these be moved to auxiliary functions?

     int string_process(...) {
             if (!process_boilerplate_pre())
                   error;
             ...
             if (!process_boilerplate_post())
                   error;
     }